### DISCUSSION TITLE
PDF Encryption after Signing a PDF

### DISCUSSION THREAD

**Discussion by TechStackGit** [MatthiasValvekens/pyHanko#188] [Q&A] [answered]
Here I wonder that after signing a PDF, I was still able to edit my PDF and after editing, my signature got invalid. How to protect/secure the PDF by encrypting it. I tried the following after signing my PDF
**` pyhanko encrypt 123.pdf 321.pdf`**
But it makes the signature invalid. What can I do about this?
Thanks in advance

**Comment by Fernando-Cabral** [comment, marked as answer]
I think what you've tried will not work because you can't change the PDF contents without making the signature invalid.
I think there are two ways you can do this (I have not checked, but it seems logical to me):
a) First encrypt, then sign - This should work.
b) After signing, encrypt as a regular file (not specifically a PDF)  - This should work too.
Since I have not tested, I can't guaranty this will work, but I think it's worth a try.

**Reply by TechStackGit** [reply]
Actually, that will work but after I encrypted the PDF I can't pass the password to pyHanko `IncrementalPdfFileWriter` module there is no option out there. So, that's why I tried to encrypt the PDF after signing it.

**Reply by TechStackGit** [reply]
When I encrypt it as a regular file the signature data are not mapped.

**Reply by MatthiasValvekens** [reply]
What @Fernando-Cabral said is essentially correct.


> Actually, that will work but after I encrypted the PDF I can't pass the password to pyHanko IncrementalPdfFileWriter module there is no option out there. So, that's why I tried to encrypt the PDF after signing it.

There is, actually: https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.pdf_utils.incremental_writer.html#pyhanko.pdf_utils.incremental_writer.IncrementalPdfFileWriter.encrypt. The naming is arguably a bit confusing, but this will do what you want (long story short, the method is named `encrypt` because the first implementation of this class started out as a patched version of `PyPDF2`'s `PdfFileWriter`).

There are quite a few tests in the test suite specifically geared towards signing encrypted files, so it's definitely possible :).

**Comment by TechStackGit** [comment]
My mistake, it worked. Thanks @Fernando-Cabral
